pkg/pyramid/tier_fs.go

```diff
@@ -311,51 +315,152 @@ func (tfs *TierFS) Open(ctx context.Context, storageID, namespace, filename stri
-				"namespace": namespace,
-				"ns_path":   nsPath,
-				"filename":  filename,
+				"storageID": fileRef.storageID,
+				"namespace": fileRef.namespace,
+				"file":      fileRef.filename,
+				"fullpath":  fileRef.fullPath,
 			}).Trace("opened locally")
 		}
 		cacheAccess.WithLabelValues(tfs.fsName, "Hit").Inc()
-		return tfs.openFile(ctx, fileRef, fh)
+		return fh, nil
 	}
 	if !os.IsNotExist(err) {
 		return nil, fmt.Errorf("open file: %w", err)
 	}
 
+	// Slow path: ensure file exists (download if needed)
 	cacheAccess.WithLabelValues(tfs.fsName, "Miss").Inc()
-	fh, err = tfs.openWithLock(ctx, fileRef)
-	if err != nil {
+	if err := tfs.ensureFileExists(ctx, fileRef); err != nil {
 		return nil, err
 	}
 
-	return tfs.openFile(ctx, fileRef, fh)
+	// Open after ensuring existence
+	fh, err = os.Open(fileRef.fullPath)
+	if err != nil {
+		return nil, fmt.Errorf("open file after download: %w", err)
+	}
+	return fh, nil
 }
 
-func (tfs *TierFS) Exists(ctx context.Context, storageID, namespace, filename string) (bool, error) {
-	cacheAccess.WithLabelValues(tfs.fsName, "Exists").Inc()
-	return tfs.adapter.Exists(ctx, tfs.objPointer(storageID, namespace, filename))
+// ensureFileExists ensures the file exists locally, downloading from block storage if needed.
+// Uses Compute to deduplicate concurrent downloads of the same file.
+func (tfs *TierFS) ensureFileExists(ctx context.Context, fileRef localFileRef) error {
+	_, err := tfs.keyLock.Compute(fileRef.filename, func() (any, error) {
+		// Check if file was downloaded by another goroutine while we waited
+		if _, err := os.Stat(fileRef.fullPath); err == nil {
+			if tfs.logger.IsTracing() {
+				tfs.log(ctx).WithFields(logging.Fields{
+					"storageID": fileRef.storageID,
+					"namespace": fileRef.namespace,
+					"file":      fileRef.filename,
+					"fullpath":  fileRef.fullPath,
+				}).Trace("got lock; file exists after all")
+			}
+			cacheAccess.WithLabelValues(tfs.fsName, "Hit").Inc()
+			return nil, nil
+		} else if !os.IsNotExist(err) {
+			return nil, fmt.Errorf("stat file: %w", err)
+		}
+
+		return nil, tfs.downloadFromStorage(ctx, fileRef)
+	})
+	return err
 }
 
-// openFile converts an os.File to pyramid.ROFile and updates the eviction control.
-func (tfs *TierFS) openFile(ctx context.Context, fileRef localFileRef, fh *os.File) (*ROFile, error) {
+// downloadFromStorage downloads a file from block storage to local disk.
+func (tfs *TierFS) downloadFromStorage(ctx context.Context, fileRef localFileRef) error {
+	if tfs.logger.IsTracing() {
+		tfs.log(ctx).WithFields(logging.Fields{
+			"storageID": fileRef.storageID,
+			"namespace": fileRef.namespace,
+			"file":      fileRef.filename,
+			"fullpath":  fileRef.fullPath,
+		}).Trace("get file from block storage")
+	}
+
+	// Use background context because Compute shares the result with multiple callers.
+	// Using a caller's context could cause one caller's cancellation to fail the fetch for all waiting callers.
```

Comment:
But please add logging fields from the calling context.  That way we have at least _one_ relevant file access on the logs.